File: 801-is-graph-bipartite/is-graph-bipartite.cpp
class Solution {
public:
    bool dfs(int start,vector<int> color,queue<int> q,vector<vector<int>>& graph){
       
       q.push(start);
       color[start]=1;
       while(!q.empty()){
        int x=q.front();
        q.pop();
        for(auto& p:graph[x]){
            if(color[p]==-1){
                q.push(p);
                color[p]=!color[x];
            }else if(color[p]==color[x]){
                return false;
            }
        }

       }
       return true;
    }
    bool isBipartite(vector<vector<int>>& graph) {
        int n=graph.size();
       vector<int> color(n,-1);
       queue<int> q;
       for(int i=0;i<n;i++){
        if(color[i]==-1){
            if(dfs(i,color,q,graph)==false){
                return false;
            }
        }
       }
       return true;
    }
};

File: 801-is-graph-bipartite/is-graph-bipartite.java
class Solution {
    public boolean isBipartite(int[][] graph) {
       int n=graph.length;
        int visited[]=new int[n];
        Arrays.fill(visited,-1);
        for(int i=0;i<n;i++){
            if(visited[i]==-1){
                visited[i]=0;
                if(!dfs(i,visited,graph)){
                    return false;
                }
            }
        }
        return true;
    }
    static boolean dfs(int node,int visited[],int[][] graph){
        for(int nd:graph[node]){
            if(visited[nd]==-1){
                if(visited[node]==1){
                    visited[nd]=0;
                }else{
                    visited[nd]=1;
                }
                if(!dfs(nd,visited,graph)){
                    return false;
                }
            }else{
                if(visited[nd]==visited[node]){
                    return false;
                }
            }
        }
        return true;
    }
}